axlearn/common/utils.py

```diff
@@ -1158,6 +1159,55 @@ def prune_tree(
     return in_tree
 
 
+def tree_merge_default_override_fn(primary: Any, secondary: Any):
+    is_primary_empty = False
+    is_secondary_empty = False
+    try:
+        is_primary_empty = len(primary) == 0
+        is_secondary_empty = len(secondary) == 0
+    except TypeError:
+        # A TypeError will be raised if primary/secondary don't have length,
+        # e.g. if they are scalars.
+        pass
+    if primary is None or is_primary_empty:
+        return secondary
+    if secondary is None or is_secondary_empty:
+        return primary
+    raise ValueError(
+        f"Encountered incompatible subtree leaves: {primary=}, {secondary=}. Specify "
+        "a custom override function to resolve incompatible subtree merges."
+    )
+
+
+def tree_merge(
+    primary: Nested[Any],
+    *,
+    secondary: Nested[Any],
+    override_fn: Callable[[Any, Any], Any] = tree_merge_default_override_fn,
+) -> Nested[Any]:
+    """Merge `secondary` into `primary`. The result contains shallow copies of subtrees from both.
```

Comment:
Consider making them deep copies to be consistent with the jax tree utils. Deep copies are also less error prone by decoupling side effects.

Reply:
Done.